### ISSUE TITLE
Add more validation to shapefile/geojson uploads

### ISSUE THREAD

**Issue by caseycesari** [WikiWatershed/model-my-watershed#1761]
- Check that coordinates are geographic
- Check that the shape is in the U.S.
- Implement a better file type check (than looking at the extension). 

It may be a good idea to create a Backbone model for uploaded data and move the validation there.

**Comment by rajadain** [comment]
Where are the shapefiles going to be stored?

**Comment by caseycesari** [comment]
Currently, and we should strive to maintain this, we are using the DataTransfer and FileReader browser APIs. This allows us to keep the file local to the user and not have to upload it anywhere.

**Comment by arottersman** [comment]
Should the guard against non-USA shapes apply to all shapes? You can currently draw AoIs anywhere you like

**Comment by arottersman** [comment]
> For shapefiles, check that the .shp and .prj are present, and that the file is otherwise valid (not sure how to do this).

What qualifies as "file is otherwise valid"? It seems like it should be enough to know if we can (1) open and parse the file (2) validate the parsed shape by our usual measures (self intersection, max-aoi, polygons, etc). We know if either of these steps fail, and I think are already  displaying errors accordingly. Is there some additional check that needs to occur?

**Comment by caseycesari** [comment]
> Should the guard against non-USA shapes apply to all shapes? You can currently draw AoIs anywhere you like

Good point. I was just thinking that we know shapes outside the US will fail, so we may want to not bother trying to analyze them, but I'll defer to @ajrobbins.

**Comment by caseycesari** [comment]
> What qualifies as "file is otherwise valid"? It seems like it should be enough to know if we can (1) open and parse the file (2) validate the parsed shape by our usual measures (self intersection, max-aoi, polygons, etc). We know if either of these steps fail, and I think are already  displaying errors accordingly. Is there some additional check that needs to occur?

Not necessarily, I just wanted to make sure we were catching and throwing error messages in all of the likely scenarios (shapefile missing components, zip is unreadable, etc). This may be something we have to iterate on as errors come up.

**Comment by ajrobbins** [comment]
>Should the guard against non-USA shapes apply to all shapes? You can currently draw AoIs anywhere you like
If we make the change here to apply to all shapes, will we use the same error modal to say why we won't analyze (with shapefile upload or regular analyze outside the US)? if the fix needs to be in two places, we should make a separate card for the general guard.

**Comment by caseycesari** [comment]
That's true. Might be worthwhile to work it in to the general workflow.
